File: Project 1/TicTacToe Java/src/TicTacToe.java
/*
 * Sam Wlody
 * CSC 242 - Project 1
 * I affirm that I have not given or received any unauthorized
 * help on this assignment, and that this work is my own.
 */

import java.util.InputMismatchException;
import java.util.Scanner;

public class TicTacToe {
	private void newGame(Scanner scan) {
		// Ask player for letter
		System.out.println("Would you like to play X or O?");
		String userLetter = scan.nextLine();
		while(!userLetter.equalsIgnoreCase("X") && !userLetter.equalsIgnoreCase("O")) {
			System.out.println("Improper input, please enter \"O\" or \"X\".");
			userLetter = scan.nextLine();
		}
		
		// Initialize to empty board
		char me;
		State currentState = new State(0);
		
		// Store character that computer will use to move, and initialize variable to keep track of turn
		me = (userLetter.equalsIgnoreCase("X") ? 'o' : 'x');
		currentState.setTurn('x');
		
		// Loop while game is not in an end state
		char winner = 'n';
		while(winner == 'n') {
			// If it's my turn to move...
			if(currentState.whoseTurn() == me) {
				// Build decision tree
				Tree decisionTree = new Tree();
				decisionTree.build(currentState);
				// Use mimimax to choose best solution
				State nextState = decisionTree.getMove(me);
				// Find difference between current state and my move
				int myMove = 0;
				for(int i=0; i<9; i++) {
					if(currentState.getBoard()[i] != nextState.getBoard()[i]) {
						myMove = i+1;
					}
				}
				// Set current state to my move and print my move
				currentState = nextState;
				System.out.println("I move to " + myMove + ".");
			} else {
				// Player move
				System.out.println("The current board is:");
				currentState.print();
				System.out.println("Where would you like to move? (1-9)");
				int playerMove = 10;
				while(true) {
					// Ensure input is valid integer, otherwise loop
					try {
						playerMove = scan.nextInt();
					} catch(InputMismatchException e) {
						System.out.println("Improper input, please enter a number from 1-9.");
						continue;
					}
					playerMove--;
					// Ensure integer is valid move and break, otherwise loop
					if(playerMove >=0 && playerMove <=8 && currentState.checkMove(playerMove)) {
						break;
					} else {
						System.out.println("Improper input, please enter a number 1-9 that does not refer to an occupied square.");
					}
				}
				// Update current state with player move
				currentState.move(playerMove);
			}
			winner = currentState.getWinner();
		}
		if(winner == 't') {
			System.out.println("The game ended in a draw! The final board state is:");
		} else {
			System.out.println("Game over, " + (winner == me ? "I" : "You") + " win! The final board state is:");
		}
		currentState.print();
	}
	
	private TicTacToe() {
		System.out.println("Welcome to TicTacToe!");
		System.out.println("Move by entering a number 1-9.");
		
		// Print sample board
		System.out.println(1 + " | " + 2 + " | " + 3);
		System.out.println("---------");
		System.out.println(4 + " | " + 5 + " | " + 6);
		System.out.println("---------");
		System.out.println(7 + " | " + 8 + " | " + 9);
		
		Scanner scan = new Scanner(System.in);
		
		String input = "Y";
		
		// Loop until user doesn't want to play any more
		while(input.equalsIgnoreCase("Y")) {
			newGame(scan); // Pass scanner in to prevent creating multiple instances
			scan.nextLine();
			System.out.println("Would you like to play again? (Y/N)");
			input = scan.nextLine();
			while(!input.equalsIgnoreCase("Y") && !input.equalsIgnoreCase("N")) {
				System.out.println("Improper input, please enter \"Y\" or \"N\".");
				input = scan.nextLine();
			}
		}
		
		System.out.println("Goodbye.");
		scan.close();
	}
	
	public static void main(String[] args) {
		new TicTacToe();
	}
}

File: Project 1/TicTacToe/TicTacToe.c
/*
 * Sam Wlody
 * CSC 242 - Project 1
 * I affirm that I have not given or received any unauthorized
 * help on this assignment, and that this work is my own.
 */

#include "State.h"
#include <stdio.h>
#include <time.h>

#define max(x,y) (x>y) ? x : y

// TODO Add some randomness to moves of equivalent value - shuffle move first?

// chars are used for most values instead of ints for smaller memory footprint and potentially faster computation
// (not sure how much of a difference this actually makes)

int states;

char alphabeta(State* state, char alpha, char beta, char color) {
    char winner = State_winner(state);
    if(winner != -1) return -winner;
    char bestValue = -127;
    int i;
    // For all possible moves, if the move is legal (not in an occupied space), explore it with negamax
    for(i=0; i<9; i++) {
        // Check move is legal (square unoccupied)
        if(state->board[i] == -1) {
            states++;
            State* explore = State_next(state, i);
            // Swap color for next call
            char v = -alphabeta(explore, -beta, -alpha, -color);
            // Free memory once state is no longer needed
            State_free(explore);
            // Set bestValue to v if v > bestValue
            bestValue = max(bestValue, v);
            alpha = max(alpha, v);
            if(alpha >= beta) break;
        }
    }
    return bestValue;
}

char getBestMove(State* state) {
    char move = -1;
    char bestValue = -127;
    char alpha = -127;
    char beta = 127;
    int i;
    for(i=0; i<9; i++) {
        if(state->board[i] == -1) {
            states++;
            State* explore = State_next(state, i);
            char v = -alphabeta(explore, -beta, -alpha, -1);
            State_free(explore);
            if(v > bestValue) {
                move = i;
                bestValue = v;
            }
            alpha = max(alpha, v);
            if(alpha >= beta) break;
        }
    }
    return move;
}

void newGame(char me) {
	// Initialize current state to empty board
    State* currentState = State_init();
    char winner = -1;
    // If State_winner() returns -1, the game is not yet over
    while(winner == -1) {
        if(currentState->player == me) {
            struct timespec start, end;
            clock_gettime(CLOCK_REALTIME, &start);
            char i = getBestMove(currentState);
            clock_gettime(CLOCK_REALTIME, &end);
            printf("%lu us\n", (end.tv_nsec - start.tv_nsec)/1000);
            printf("%d states explored.\n", states);
            printf("%s%d%c\n", "I move to space ", i+1, '.');
            currentState = State_next(currentState, i);
        } else {
            State_print(currentState);
            printf("Where would you like to move?\n");
            int i;
            // Ensure an integer is entered, then ensure integer is between 1 and 9, then ensure move is legal (space is empty)
            while(scanf(" %d", &i) == EOF || i<1 || i>9 || currentState->board[i-1] != -1) {
                printf("Improper input, please enter a number 1-9 that corresponds to an empty space.\n");
            }
            currentState = State_next(currentState, i-1);
        }
        // Check if game is over
        winner = State_winner(currentState);
    }
    if(winner) {
        // Whoever's turn it is lost
        printf("%s won!\n", currentState->player == me ? "You" : "I");
    } else {
        printf("Draw!\n");
    }
}

int main(int argc, char* argv[]) {
    printf("Welcome to TicTacToe!\n");
    printf("Move by entering a number 1-9.\n");
    char input = 'y';
    while(input == 'y' || input == 'Y') {
        printf("Would you like to play X or O?\n");
        char player;
        // Ensure a character is entered, then ensure it is equal to x or y
        while(scanf(" %c", &player) == EOF || (player != 'x' && player != 'X' && player != 'o' && player != 'O')) {
            printf("Improper input, please enter either \"X\" or \"O\".\n");
        }
        newGame(player == 'x' || player == 'X');
        printf("Would you like to play again? (Y/N)\n");
        // Ensure a character is entered, then ensure it is equal to y or n
        while(scanf(" %c", &input) == EOF || (input != 'y' && input != 'Y' && input != 'n' && input != 'N')) {
            printf("Improper input, please enter either \"Y\" or \"N\".\n");
        }
    }
    printf("Goodbye.\n");
    return 0;
}
